### PR TITLE
Support custom trainer and backend

### PR DESCRIPTION
**Pull request by KKIEEK**
### Motivation
Since MM-based repositories already use `torch.distributed.init_process_group`, using `TorchTrainer` for DDP in ray framework causes the `RuntimeError("trying to initialize the default process group twice!")`.
To solve this problem, I introduced a custom backend modified from [here](https://github.com/ray-project/ray/blob/ray-2.1.0/python/ray/train/torch/config.py).

### CODE REVIEW

File: siatune/ray/config.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            'runtime environment: '
+            "`ray.init(runtime_env={{'env_vars': {'NCCL_SOCKET_IFNAME': 'ens5'}})`"  # noqa
+        )
+        os.environ['NCCL_SOCKET_IFNAME'] = DEFAULT_NCCL_SOCKET_IFNAME
+
+
+class _CustomTorchBackend(_TorchBackend):
+    share_cuda_visible_devices: bool = True
+
+    def on_start(self, worker_group: WorkerGroup,
```
Comment (yhna940) [inline siatune/ray/config.py:46]: If the process group is not initialized, how about initializing it here without throwing an error?

**Review by yhna940 — APPROVED** [review]

File: siatune/ray/config.py (hunk 1 above)
Comment (KKIEEK) [reply siatune/ray/config.py:46]: I think we have no way to know whether the process group is initialized in `Task.run` method.

File: siatune/ray/config.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            'to prioritize ethernet connection. To override this behavior, '
+            'set the `NCCL_SOCKET_IFNAME` environment variable in your Ray '
+            'runtime environment: '
+            "`ray.init(runtime_env={{'env_vars': {'NCCL_SOCKET_IFNAME': 'ens5'}})`"  # noqa
+        )
+        os.environ['NCCL_SOCKET_IFNAME'] = DEFAULT_NCCL_SOCKET_IFNAME
+
+
+class _CustomTorchBackend(_TorchBackend):
+    share_cuda_visible_devices: bool = True
```
Comment (nijkah) [inline siatune/ray/config.py:44]:
```suggestion
```

File: siatune/ray/config.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    if 'NCCL_SOCKET_IFNAME' not in os.environ:
+        logger.debug(
+            f'Setting NCCL_SOCKET_IFNAME to {DEFAULT_NCCL_SOCKET_IFNAME} '
+            'to prioritize ethernet connection. To override this behavior, '
+            'set the `NCCL_SOCKET_IFNAME` environment variable in your Ray '
+            'runtime environment: '
+            "`ray.init(runtime_env={{'env_vars': {'NCCL_SOCKET_IFNAME': 'ens5'}})`"  # noqa
+        )
+        os.environ['NCCL_SOCKET_IFNAME'] = DEFAULT_NCCL_SOCKET_IFNAME
```
[the remark this answers is not in the document]
Comment (KKIEEK) [reply siatune/ray/config.py]: I removed these lines.

**Review by nijkah — APPROVED** [review]

File: siatune/ray/config.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) SI-Analytics. All rights reserved.
+# Modified from https://github.com/ray-project/ray/blob/ray-2.1.0/python/ray/train/torch/config.py  # noqa
+# for applying MM based repo training
```
Comment (KKIEEK) [inline siatune/ray/config.py:3]:
```suggestion
```

**Pull request merged**